File: example-app/src/main/java/com/adyen/example/model/request/ShopperName.java
/*
 * Copyright (c) 2018 Adyen N.V.
 *
 * This file is open source and available under the MIT license. See the LICENSE file for more info.
 *
 * Created by caiof on 16/11/2018.
 */

package com.adyen.example.model.request;

import com.squareup.moshi.Json;

import java.io.Serializable;

public class ShopperName implements Serializable {

    @Json(name = "firstName")
    private String mFirstName;

    @Json(name = "infix")
    private String mInfix;

    @Json(name = "lastName")
    private String mLastName;

    @Json(name = "gender")
    private Gender mGender;


    public void setFirstName(String firstName) {
        this.mFirstName = firstName;
    }

    public void setInfix(String infix) {
        this.mInfix = infix;
    }

    public void setLastName(String lastName) {
        this.mLastName = lastName;
    }

    public void setGender(Gender gender) {
        this.mGender = gender;
    }


    public enum Gender {
        @Json(name = "M")
        MALE,
        @Json(name = "F")
        FEMALE
    }

}

File: core/src/main/java/com/adyen/checkout/core/components/data/ShopperName.kt
/*
 * Copyright (c) 2025 Adyen N.V.
 *
 * This file is open source and available under the MIT license. See the LICENSE file for more info.
 *
 * Created by ozgur on 12/5/2025.
 */

package com.adyen.checkout.core.components.data

import com.adyen.checkout.core.common.exception.ModelSerializationException
import com.adyen.checkout.core.common.internal.model.ModelObject
import com.adyen.checkout.core.common.internal.model.getStringOrNull
import kotlinx.parcelize.Parcelize
import org.json.JSONException
import org.json.JSONObject

@Parcelize
data class ShopperName(
    var firstName: String? = null,
    var infix: String? = null,
    var lastName: String? = null,
    var gender: String? = null,
) : ModelObject() {

    companion object {
        private const val FIRST_NAME = "firstName"
        private const val INFIX = "infix"
        private const val LAST_NAME = "lastName"
        private const val GENDER = "gender"

        @JvmField
        val SERIALIZER: Serializer<ShopperName> = object : Serializer<ShopperName> {
            override fun serialize(modelObject: ShopperName): JSONObject {
                return try {
                    JSONObject().apply {
                        putOpt(FIRST_NAME, modelObject.firstName)
                        putOpt(INFIX, modelObject.infix)
                        putOpt(LAST_NAME, modelObject.lastName)
                        putOpt(GENDER, modelObject.gender)
                    }
                } catch (e: JSONException) {
                    throw ModelSerializationException(ShopperName::class.java, e)
                }
            }

            override fun deserialize(jsonObject: JSONObject): ShopperName {
                return ShopperName(
                    firstName = jsonObject.getStringOrNull(FIRST_NAME),
                    infix = jsonObject.getStringOrNull(INFIX),
                    lastName = jsonObject.getStringOrNull(LAST_NAME),
                    gender = jsonObject.getStringOrNull(GENDER),
                )
            }
        }
    }
}
